src/Processors/Formats/Impl/NativeORCBlockInputFormat.cpp

```diff
@@ -305,15 +407,30 @@ static void buildORCSearchArgumentImpl(
                 }
             }
 
-            const auto * type
-                = getORCTypeByName(schema, getColumnNameFromKeyCondition(key_condition, curr.key_column), case_insensitive_column_matching);
-            if (!type)
+            String column_name = getColumnNameFromKeyCondition(key_condition, curr.key_column);
+            const auto * orc_type = getORCTypeByName(schema, column_name, format_settings.orc.case_insensitive_column_matching);
+            if (!orc_type)
             {
                 builder.literal(orc::TruthValue::YES_NO_NULL);
                 break;
             }
 
-            auto predicate_type = convertORCTypeToPredicateType(*type);
+            /// Make sure key column in header has exactly the same type with key column in ORC file schema
+            /// A counter-example:
+            ///     Column a has type "Nullable(Int64)" in ORC file, but in header column a has type "Int64", which is allowd in CH.
+            ///     For queries with where condition like "a is null", pushing such filter down to ORC file will make CH return wrong results.
```

Comment:
Column can also be requested as nullable/non-nullable by explicitly specifying the schema: `select * from file('t.orc', ORC, 'x Int64')` , independently of the `schema_inference_make_columns_nullable` setting. I'd guess that this is used often enough that we want filter pushdown to work in this case.

I think the logic should be: if the column is non-nullable in the header, but nullable in the file, and `format_settings.null_as_default` is true - then we're in trouble.  Otherwise we're fine. If we're in trouble, we can do YES_NO_NULL for now, but ideally we want to handle this case too, because it seems common [1]. The null-as-default can be handled by checking if the default value passes the FUNCTION_[NOT_]IN_RANGE/IN_SET/NULL condition and adding either "or is null" or "and is not null" to the orc search argument, I guess.

[1] `input_format_null_as_default` is enabled by default; columns in the file are nullable if written by clickhouse using INSERT SELECT from another orc/parquet file; and specifying schema explicitly seems common).

Reply:
schema_inference_make_columns_nullable is removed now. Because `expect_type` is the actual column type in orc files. It should always be wrapped with nullable.

Reply:
The difference of `Nullable(T)` in orc file and `T` in format header is just one of the corner cases we should consider.

Another corner case, for example, `Nullable(Int64)` in orc file, and `Nullable(UInt64)` in format header, and filter expr: `a > 100`.  If we don't process it,  value `-1` in column a would pass the filter.

Reply:
> Another corner case, for example, Nullable(Int64) in orc file, and Nullable(UInt64) in format header

Ah, right, I forgot we allow that. My uninformed guess is that people do this much less often than the Nullable(T) -> T thing, so supporting only the Nullable thing sounds reasonable.